### ISSUE TITLE
Extend allowed IP's to entire subnets for rapidly changing mobile environments

### ISSUE THREAD

**Issue by mrash** [mrash/fwknop#138]
There have been a couple of reports (from Gerry Reno and also from the blog post below) of fwknop usage in mobile networks where the client IP changes across very short time scales so the IP to be allowed at the time of fwknop client usage is invalid by the time the actual service is used. Here is an example:

http://steveshomelab.wordpress.com/2014/10/29/mobile-internet-ip-address-craziness-fun-with-single-packet-authorization/

The SPA format should be extended to allow entire subnets through the firewall where fwknopd is running. Although this is not a perfect solution, it should help somewhat in scenarios like this (imagine using a class A if necessary).

**Comment by mrash** [comment]
This can be implemented without changing the SPA packet format by using a server-side directive. At least this could be an option to easily offer the feature, and there is precedent for things like this with the FORCE_* vars.

**Comment by oneru** [comment]
I wonder if there is actually something else going on with this bug.  It almost looks like the mobile carrier is running the https connections through a set of proxies that are load-balancing the connections.  

I'm not against adding this option to fwknopd, but it might not even solve the problem, as the proxies are not necessarily on the same subnet as the real IP of the mobile device.

I'm speculating, of course, but I think this is the most likely explanation.

**Comment by mrash** [comment]
@oneru Yes, that is a good point. I suspect this won't solve the problem completely (or even "mostly") in such environments. But, perhaps opening a service to any IP and then using the idea from @damienstuart about shutting down access as soon as the first connection is made would make for a reasonable solution. If the client can't know from where the actual connection will emanate because of a proxy / load balancing situation, this may be the only option.

**Comment by oneru** [comment]
@mrash Another approach to solving this would be an attempt to determine the device's assigned IP address from the network adapter itself.   This approach will take effort to get it working reliably.  It is, however, something that will be even more valuable with ipv6, as nat routing will presumably vanish in that environment.

This would apply most to the android client right now.  Ideally, we would grab the IP address of the network device that has the default route.  If this is a routable IP, hence not behind NAT, use that as the source IP.  If not, either fail, or fail-over to another method.  This could be added to the other clients, but desktop clients would be much more likely to sit behind nat routing, at least in the context of ipv4.  Once I'm back at my Android development machine, I'll see if I can get this working.

**Comment by mrash** [comment]
Although having the IP assigned to the network adapter would be useful as an option, I'm not sure a routable IP would necessary imply no NAT / load balancing / proxies upstream. In the original bug report and the example mentioned at the top of this issue, I don't think it was specified whether the local IP on the device itself was rapidly changing - just that an external IP resolution service showed that, from its perspective, the IP was changing. Another issue is that even if there always is a 1:1 correspondence between the local IP and what external systems would see, there would be a race condition between getting the local IP, placing it within a SPA packet, and then making the follow-on SSH (or whatever) connection. This race would be similar to the current race when an external IP resolution service is used, but the local one would be a lot faster.

I think it would be good to have both approaches to account for however the mobile carrier network is built. For those where the local IP is always a reliable indicator of the external IP, then your solution would work subject to the race.

If there were any way to get actual evidence for various provider networks that would be great, but probably somewhat hard.

**Comment by oneru** [comment]
OK, my assumption that cell companies are well behaved was unfounded.  On my T-Mobile phone, the network device gets a 33.x.x.x address, which is a routable address assigned to the DoD (not spooky at all), but the outside world sees my IP as a 172.56.x.x address.  A traceroute launched from my device shows packets bounce around in a 10.x.x.x address space before emerging on a Level3 network in Detroit.  So T-Mobile is using carrier grade nat, but is squatting on the 33.x network, and using it as a not publically routable network.  This means that pulling an IP from the network device is a no-go strategy.

**Comment by leonyu** [comment]
This would be useful for my setup. I am using fwknop client exclusively on my phone as a "two-factor auth", while the actual credentials are on my various computers. 

Currently i have profiles for each source-destination combo I use, but I need to edit them because of dynamic ips. This feature would alleviate that soI don't need to edit them as often.

**Comment by jp-bennett** [comment]
@leonyu, you might be interested in a service I host: https://qr.incom.co

Assuming you're using the Android app, in the config, change The "allow ip" option to "prompt IP".  When the prompt comes up, there should be an option to scan qr.  The scan qr allows you to just point the phone at the qr code.  It contains the IP of the computer, which gets embedded in the SPA packet.

**Comment by leonyu** [comment]
Ah, that would work for all the desktops in this setup. I do have 2 "jump hosts" that are terminal only, which that wouldn't work for, tho only one of them is dynamic IP.

**Comment by jp-bennett** [comment]
@leonyu You've just given me an interesting idea.  Just for kicks, take a look at this command: 

curl https://incomsystems.biz/fwknop-gui/qr-txt.php

**Comment by leonyu** [comment]
oh wow, didn't think of using a terminal application to generate QR code. 

Your script works ! 

Thought as a privacy conscious person, I didn't want to hit your server every time I need to auth myself, so I looked into the available commands a bit. There is `qrencode` which is available in most distros, and made a one-liner: 

```sh
qrencode -t ANSI "CLIENT_IP:$(curl https://domains.google.com/checkip)"
```

**Comment by jp-bennett** [comment]
@leonyu I use qrencode on the qr.incom.co server, and that also powers the qr code generation in the openwrt/lede interface. :)
